File: examples/ch15/comparison.c
/**
 * C语言安全编程对比示例
 * 
 * 本文件展示C语言中的安全编程实践，包括：
 * 1. 输入验证和消毒
 * 2. 缓冲区溢出防护
 * 3. 安全字符串处理
 * 4. 内存安全
 * 5. 文件操作安全
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdbool.h>
#include <time.h>

/* 最大缓冲区大小 */
#define MAX_BUFFER_SIZE 1024
#define MAX_PATH_LENGTH 260
#define MAX_FILENAME_LENGTH 255

/* 危险字符列表 */
#define DANGEROUS_CHARS "\\:*?\"<>|&;`$"

/**
 * 验证用户输入是否安全
 * @param input 用户输入
 * @param max_length 最大长度
 * @return 是否安全
 */
bool validate_input(const char *input, size_t max_length) {
    /* 检查空指针 */
    if (input == NULL) {
        return false;
    }
    
    /* 检查长度 */
    if (strlen(input) > max_length) {
        return false;
    }
    
    /* 检查空字符串 */
    if (strlen(input) == 0) {
        return false;
    }
    
    /* 检查危险字符 */
    for (size_t i = 0; i < strlen(input); i++) {
        if (strchr(DANGEROUS_CHARS, input[i]) != NULL) {
            return false;
        }
    }
    
    /* 检查路径遍历 */
    if (strstr(input, "..") != NULL) {
        return false;
    }
    
    return true;
}

/**
 * 验证文件名是否安全
 * @param filename 文件名
 * @return 是否安全
 */
bool validate_filename(const char *filename) {
    /* 检查空指针 */
    if (filename == NULL) {
        return false;
    }
    
    /* 检查空字符串 */
    if (strlen(filename) == 0) {
        return false;
    }
    
    /* 检查长度 */
    if (strlen(filename) > MAX_FILENAME_LENGTH) {
        return false;
    }
    
    /* 检查非法字符 */
    const char *illegal_chars = "\\:*?\"<>|";
    for (size_t i = 0; i < strlen(filename); i++) {
        if (strchr(illegal_chars, filename[i]) != NULL) {
            return false;
        }
    }
    
    return true;
}

/**
 * 验证路径是否安全
 * @param path_input 路径
 * @return 是否安全
 */
bool validate_path(const char *path_input) {
    /* 检查空指针 */
    if (path_input == NULL) {
        return false;
    }
    
    /* 检查空字符串 */
    if (strlen(path_input) == 0) {
        return false;
    }
    
    /* 检查路径遍历 */
    if (strstr(path_input, "..") != NULL) {
        return false;
    }
    
    /* 检查危险字符 */
    const char *dangerous_chars = "&|<>^";
    for (size_t i = 0; i < strlen(path_input); i++) {
        if (strchr(dangerous_chars, path_input[i]) != NULL) {
            return false;
        }
    }
    
    return true;
}

/**
 * 安全字符串复制
 * @param dest 目标缓冲区
 * @param src 源字符串
 * @param dest_size 目标缓冲区大小
 * @return 是否成功
 */
bool safe_strcpy(char *dest, const char *src, size_t dest_size) {
    /* 检查空指针 */
    if (dest == NULL || src == NULL || dest_size == 0) {
        return false;
    }
    
    size_t src_len = strlen(src);
    
    /* 检查是否需要截断 */
    if (src_len >= dest_size) {
        /* 截断字符串 */
        strncpy(dest, src, dest_size - 1);
        dest[dest_size - 1] = '\0';
        return false;
    }
    
    /* 安全复制 */
    strcpy(dest, src);
    return true;
}

/**
 * 安全字符串连接
 * @param dest 目标缓冲区
 * @param src 源字符串
 * @param dest_size 目标缓冲区大小
 * @return 是否成功
 */
bool safe_strcat(char *dest, const char *src, size_t dest_size) {
    /* 检查空指针 */
    if (dest == NULL || src == NULL || dest_size == 0) {
        return false;
    }
    
    size_t dest_len = strlen(dest);
    size_t src_len = strlen(src);
    
    /* 检查是否有足够空间 */
    if (dest_len + src_len >= dest_size) {
        /* 截断连接 */
        size_t available = dest_size - dest_len - 1;
        strncat(dest, src, available);
        return false;
    }
    
    /* 安全连接 */
    strcat(dest, src);
    return true;
}

/**
 * 安全格式化字符串
 * @param buffer 输出缓冲区
 * @param size 缓冲区大小
 * @param format 格式化字符串
 * @param ... 可变参数
 * @return 写入的字符数，失败返回-1
 */
int safe_snprintf(char *buffer, size_t size, const char *format, ...) {
    /* 检查参数 */
    if (buffer == NULL || size == 0 || format == NULL) {
        return -1;
    }
    
    va_list args;
    va_start(args, format);
    
    int result = vsnprintf(buffer, size, format, args);
    
    va_end(args);
    
    /* 检查是否截断 */
    if (result >= (int)size) {
        buffer[size - 1] = '\0';
        return -1;
    }
    
    return result;
}

/**
 * 消毒用户输入
 * @param input 原始输入
 * @param output 输出缓冲区
 * @param output_size 输出缓冲区大小
 * @return 是否成功
 */
bool sanitize_input(const char *input, char *output, size_t output_size) {
    /* 检查参数 */
    if (input == NULL || output == NULL || output_size == 0) {
        return false;
    }
    
    size_t j = 0;
    size_t input_len = strlen(input);
    
    /* 遍历输入字符串 */
    for (size_t i = 0; i < input_len && j < output_size - 1; i++) {
        char c = input[i];
        
        /* 跳过危险字符 */
        if (strchr(DANGEROUS_CHARS, c) != NULL) {
            continue;
        }
        
        /* 跳过路径遍历 */
        if (c == '.' && i + 1 < input_len && input[i + 1] == '.') {
            i++; /* 跳过第二个点 */
            continue;
        }
        
        /* 添加安全字符 */
        output[j++] = c;
    }
    
    /* 添加字符串结束符 */
    output[j] = '\0';
    
    return true;
}

/**
 * 消毒文件名
 * @param filename 原始文件名
 * @param output 输出缓冲区
 * @param output_size 输出缓冲区大小
 * @return 是否成功
 */
bool sanitize_filename(const char *filename, char *output, size_t output_size) {
    /* 检查参数 */
    if (filename == NULL || output == NULL || output_size == 0) {
        return false;
    }
    
    size_t j = 0;
    size_t filename_len = strlen(filename);
    const char *illegal_chars = "\\:*?\"<>|";
    
    /* 遍历文件名 */
    for (size_t i = 0; i < filename_len && j < output_size - 1; i++) {
        char c = filename[i];
        
        /* 跳过非法字符 */
        if (strchr(illegal_chars, c) != NULL) {
            continue;
        }
        
        /* 添加安全字符 */
        output[j++] = c;
    }
    
    /* 添加字符串结束符 */
    output[j] = '\0';
    
    return true;
}

/**
 * 安全路径连接
 * @param base_path 基础路径
 * @param user_path 用户路径
 * @param output 输出缓冲区
 * @param output_size 输出缓冲区大小
 * @return 是否安全
 */
bool safe_path_join(const char *base_path, const char *user_path, 
                   char *output, size_t output_size) {
    /* 检查参数 */
    if (base_path == NULL || user_path == NULL || 
        output == NULL || output_size == 0) {
        return false;
    }
    
    /* 验证用户路径 */
    if (!validate_path(user_path)) {
        return false;
    }
    
    /* 检查用户路径是否为绝对路径 */
    if (strlen(user_path) >= 2 && user_path[1] == ':') {
        return false;
    }
    
    /* 构建完整路径 */
    safe_strcpy(output, base_path, output_size);
    safe_strcat(output, "\\", output_size);
    safe_strcat(output, user_path, output_size);
    
    return true;
}

/**
 * 安全文件读取
 * @param file_path 文件路径
 * @param buffer 输出缓冲区
 * @param buffer_size 缓冲区大小
 * @return 是否成功
 */
bool safe_file_read(const char *file_path, char *buffer, size_t buffer_size) {
    /* 检查参数 */
    if (file_path == NULL || buffer == NULL || buffer_size == 0) {
        return false;
    }
    
    /* 打开文件 */
    FILE *file = fopen(file_path, "r");
    if (file == NULL) {
        return false;
    }
    
    /* 读取文件内容 */
    size_t bytes_read = fread(buffer, 1, buffer_size - 1, file);
    buffer[bytes_read] = '\0';
    
    /* 关闭文件 */
    fclose(file);
    
    return true;
}

/**
 * 安全文件写入
 * @param file_path 文件路径
 * @param content 文件内容
 * @return 是否成功
 */
bool safe_file_write(const char *file_path, const char *content) {
    /* 检查参数 */
    if (file_path == NULL || content == NULL) {
        return false;
    }
    
    /* 打开文件 */
    FILE *file = fopen(file_path, "w");
    if (file == NULL) {
        return false;
    }
    
    /* 写入内容 */
    size_t content_len = strlen(content);
    size_t bytes_written = fwrite(content, 1, content_len, file);
    
    /* 关闭文件 */
    fclose(file);
    
    return bytes_written == content_len;
}

/**
 * 简单的哈希函数（仅用于演示）
 * @param input 输入字符串
 * @return 哈希值
 */
unsigned long simple_hash(const char *input) {
    unsigned long hash = 5381;
    int c;
    
    while ((c = *input++)) {
        hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
    }
    
    return hash;
}

/**
 * 演示输入验证
 */
void demonstrate_input_validation(void) {
    printf("============================================================\n");
    printf("输入验证演示\n");
    printf("============================================================\n");
    
    /* 测试用例 */
    struct {
        const char *input;
        bool expected;
        const char *description;
    } test_cases[] = {
        {"", false, "空输入"},
        {"hello_world", true, "正常输入"},
        {"file.txt & del /f /q C:\\*.*", false, "命令注入攻击"},
        {"..\\..\\..\\Windows\\System32\\config\\SAM", false, "路径遍历攻击"},
        {"file:name.txt", false, "包含非法字符"},
        {"normal_file.txt", true, "正常文件名"},
        {NULL, false, NULL}
    };
    
    for (int i = 0; test_cases[i].input != NULL; i++) {
        bool result = validate_input(test_cases[i].input, MAX_BUFFER_SIZE);
        const char *status = (result == test_cases[i].expected) ? "PASS" : "FAIL";
        printf("[%s] %s: '%s' -> %s\n", 
               status, test_cases[i].description, 
               test_cases[i].input, result ? "true" : "false");
    }
}

/**
 * 演示路径安全
 */
void demonstrate_path_security(void) {
    printf("\n============================================================\n");
    printf("路径安全演示\n");
    printf("============================================================\n");
    
    /* 测试用例 */
    struct {
        const char *input;
        bool expected;
        const char *description;
    } test_cases[] = {
        {"file.txt", true, "正常文件"},
        {"..\\..\\secret.txt", false, "路径遍历攻击"},
        {"C:\\Windows\\System32\\cmd.exe", false, "绝对路径"},
        {"subdir\\file.txt", true, "子目录文件"},
        {NULL, false, NULL}
    };
    
    char base_path[] = "C:\\Users\\test";
    char output[MAX_PATH_LENGTH];
    
    for (int i = 0; test_cases[i].input != NULL; i++) {
        bool result = safe_path_join(base_path, test_cases[i].input, 
                                    output, sizeof(output));
        const char *status = (result == test_cases[i].expected) ? "PASS" : "FAIL";
        printf("[%s] %s: '%s' -> %s\n", 
               status, test_cases[i].description, 
               test_cases[i].input, result ? "安全" : "不安全");
    }
}

/**
 * 演示字符串安全
 */
void demonstrate_string_security(void) {
    printf("\n============================================================\n");
    printf("字符串安全演示\n");
    printf("============================================================\n");
    
    /* 测试安全字符串复制 */
    char buffer[20];
    const char *long_string = "This is a very long string that should be truncated";
    
    printf("安全字符串复制:\n");
    printf("  原始字符串: '%s'\n", long_string);
    printf("  缓冲区大小: %zu\n", sizeof(buffer));
    
    bool result = safe_strcpy(buffer, long_string, sizeof(buffer));
    printf("  复制结果: %s\n", result ? "成功" : "截断");
    printf("  复制后字符串: '%s'\n", buffer);
    
    /* 测试安全字符串连接 */
    printf("\n安全字符串连接:\n");
    char dest[30] = "Hello, ";
    const char *src = "World! This is a test.";
    
    printf("  目标字符串: '%s'\n", dest);
    printf("  源字符串: '%s'\n", src);
    
    result = safe_strcat(dest, src, sizeof(dest));
    printf("  连接结果: %s\n", result ? "成功" : "截断");
    printf("  连接后字符串: '%s'\n", dest);
}

/**
 * 演示输入消毒
 */
void demonstrate_sanitization(void) {
    printf("\n============================================================\n");
    printf("输入消毒演示\n");
    printf("============================================================\n");
    
    /* 测试用例 */
    struct {
        const char *input;
        const char *description;
    } test_cases[] = {
        {"file:name.txt", "文件名消毒"},
        {"  hello world  ", "空格处理"},
        {"file.txt & command", "命令注入消毒"},
        {"..\\..\\path", "路径遍历消毒"},
        {NULL, NULL}
    };
    
    char sanitized[MAX_BUFFER_SIZE];
    
    for (int i = 0; test_cases[i].input != NULL; i++) {
        printf("%s:\n", test_cases[i].description);
        printf("  输入: '%s'\n", test_cases[i].input);
        
        sanitize_input(test_cases[i].input, sanitized, sizeof(sanitized));
        printf("  输出: '%s'\n", sanitized);
        printf("\n");
    }
}

/**
 * 演示文件操作安全
 */
void demonstrate_file_security(void) {
    printf("\n============================================================\n");
    printf("文件操作安全演示\n");
    printf("============================================================\n");
    
    /* 创建测试文件 */
    const char *test_file = "test_security.txt";
    const char *test_content = "Hello, this is a test file for security demonstration.";
    
    printf("创建测试文件: %s\n", test_file);
    if (safe_file_write(test_file, test_content)) {
        printf("  文件创建成功\n");
        
        /* 读取文件 */
        char buffer[MAX_BUFFER_SIZE];
        if (safe_file_read(test_file, buffer, sizeof(buffer))) {
            printf("  文件内容: '%s'\n", buffer);
        } else {
            printf("  文件读取失败\n");
        }
        
        /* 计算哈希 */
        unsigned long hash = simple_hash(buffer);
        printf("  文件哈希: %lu\n", hash);
        
        /* 删除测试文件 */
        remove(test_file);
        printf("  测试文件已删除\n");
    } else {
        printf("  文件创建失败\n");
    }
}

/**
 * 主函数
 */
int main(void) {
    printf("C语言安全编程对比示例\n");
    printf("============================================================\n");
    
    demonstrate_input_validation();
    demonstrate_path_security();
    demonstrate_string_security();
    demonstrate_sanitization();
    demonstrate_file_security();
    
    printf("\n============================================================\n");
    printf("所有演示完成\n");
    printf("============================================================\n");
    
    return 0;
}

File: examples/ch15/comparison.lua
--[[
    Lua安全编程对比示例
    
    本文件展示Lua中的安全编程实践，包括：
    1. 输入验证和消毒
    2. 路径遍历防护
    3. 安全文件操作
    4. 敏感信息处理
    5. 日志记录
    6. 哈希计算
]]

-- 危险字符列表
local DANGEROUS_CHARS = "[\\/:*?\"<>|&;`$]"

-- 危险模式列表
local DANGEROUS_PATTERNS = {
    "%.%.",  -- 路径遍历
    "[&|<>^]",  -- 命令注入
    "[;]",  -- 命令分隔符
}

-- 非法文件名字符
local ILLEGAL_FILENAME_CHARS = "[\\/:*?\"<>|]"

-- ============================================================
-- 安全验证器
-- ============================================================

local SecurityValidator = {}

function SecurityValidator.validate_input(user_input)
    -- 验证用户输入是否安全
    
    -- 检查空输入
    if user_input == nil or user_input == "" then
        return false
    end
    
    -- 检查危险字符
    if string.find(user_input, DANGEROUS_CHARS) then
        return false
    end
    
    -- 检查危险模式
    for _, pattern in ipairs(DANGEROUS_PATTERNS) do
        if string.find(user_input, pattern) then
            return false
        end
    end
    
    return true
end

function SecurityValidator.validate_filename(filename)
    -- 验证文件名是否安全
    
    -- 检查空文件名
    if filename == nil or filename == "" then
        return false
    end
    
    -- 检查非法字符
    if string.find(filename, ILLEGAL_FILENAME_CHARS) then
        return false
    end
    
    -- 检查长度
    if #filename > 255 then
        return false
    end
    
    return true
end

function SecurityValidator.validate_path(path_input)
    -- 验证路径是否安全
    
    -- 检查空路径
    if path_input == nil or path_input == "" then
        return false
    end
    
    -- 检查路径遍历
    if string.find(path_input, "%.%.") then
        return false
    end
    
    -- 检查危险字符
    if string.find(path_input, "[&|<>^]") then
        return false
    end
    
    return true
end

-- ============================================================
-- 安全路径处理器
-- ============================================================

local SafePathHandler = {}
SafePathHandler.__index = SafePathHandler

function SafePathHandler.new(base_dir)
    -- 创建安全路径处理器
    
    local self = setmetatable({}, SafePathHandler)
    self.base_dir = base_dir
    
    -- 确保基础目录存在
    os.execute(string.format('mkdir "%s" 2>nul', base_dir))
    
    return self
end

function SafePathHandler:safe_join(user_path)
    -- 安全地连接路径
    
    -- 验证用户路径
    if not SecurityValidator.validate_path(user_path) then
        return nil
    end
    
    -- 检查是否为绝对路径
    if string.match(user_path, "^[A-Za-z]:") then
        return nil
    end
    
    -- 构建完整路径
    local full_path = self.base_dir .. "\\" .. user_path
    
    -- 规范化路径
    full_path = string.gsub(full_path, "\\", "/")
    full_path = string.gsub(full_path, "/+", "/")
    
    -- 验证路径是否在基础目录内
    local normalized_base = string.gsub(self.base_dir, "\\", "/")
    if not string.find(full_path, normalized_base, 1, true) then
        return nil
    end
    
    return full_path
end

function SafePathHandler:safe_read(filename)
    -- 安全地读取文件
    
    -- 验证路径
    local file_path = self:safe_join(filename)
    if file_path == nil then
        return nil
    end
    
    -- 检查文件是否存在
    local file = io.open(file_path, "r")
    if file == nil then
        return nil
    end
    
    -- 读取文件内容
    local content = file:read("*a")
    file:close()
    
    return content
end

function SafePathHandler:safe_write(filename, content)
    -- 安全地写入文件
    
    -- 验证路径
    local file_path = self:safe_join(filename)
    if file_path == nil then
        return false
    end
    
    -- 写入文件
    local file = io.open(file_path, "w")
    if file == nil then
        return false
    end
    
    file:write(content)
    file:close()
    
    return true
end

-- ============================================================
-- 安全哈希计算器
-- ============================================================

local SecureHasher = {}

function SecureHasher.calculate_simple_hash(input)
    -- 计算简单的哈希值（仅用于演示）
    
    local hash = 5381
    for i = 1, #input do
        hash = ((hash * 33) + string.byte(input, i)) % 1000000007
    end
    
    return hash
end

function SecureHasher.hash_password(password, salt)
    -- 安全地哈希密码（简化版本）
    
    if salt == nil then
        -- 生成随机盐
        math.randomseed(os.time())
        salt = ""
        for i = 1, 32 do
            salt = salt .. string.char(math.random(65, 122))
        end
    end
    
    -- 简单哈希（实际应用应使用更强的哈希函数）
    local combined = password .. salt
    local hash = SecureHasher.calculate_simple_hash(combined)
    
    return {
        salt = salt,
        hash = tostring(hash)
    }
end

-- ============================================================
-- 审计日志记录器
-- ============================================================

local AuditLogger = {}
AuditLogger.__index = AuditLogger

function AuditLogger.new(log_dir)
    -- 创建审计日志记录器
    
    local self = setmetatable({}, AuditLogger)
    self.log_dir = log_dir
    self.log_file = log_dir .. "\\audit.log"
    
    -- 确保日志目录存在
    os.execute(string.format('mkdir "%s" 2>nul', log_dir))
    
    return self
end

function AuditLogger:log_event(event_type, message, user)
    -- 记录审计事件
    
    if user == nil then
        user = os.getenv("USERNAME") or "unknown"
    end
    
    local timestamp = os.date("%Y-%m-%d %H:%M:%S")
    local computer = os.getenv("COMPUTERNAME") or "unknown"
    local log_message = string.format("[%s] [%s] User: %s Computer: %s - %s\n", 
                                      timestamp, event_type, user, computer, message)
    
    local file = io.open(self.log_file, "a")
    if file then
        file:write(log_message)
        file:close()
    end
    
    -- 同时输出到控制台
    print(string.format("[AUDIT] %s", log_message))
end

function AuditLogger:log_security_event(event_type, message)
    -- 记录安全事件
    
    self:log_event("SECURITY_" .. event_type, message)
end

function AuditLogger:log_error(error_code, message)
    -- 记录错误事件
    
    local timestamp = os.date("%Y-%m-%d %H:%M:%S")
    local log_message = string.format("[%s] [ERROR] Code: %s - %s\n", 
                                      timestamp, error_code, message)
    
    local file = io.open(self.log_file, "a")
    if file then
        file:write(log_message)
        file:close()
    end
    
    print(string.format("[ERROR] %s", log_message))
end

-- ============================================================
-- 输入消毒器
-- ============================================================

local InputSanitizer = {}

function InputSanitizer.sanitize_filename(filename)
    -- 消毒文件名
    
    -- 移除非法字符
    local sanitized = string.gsub(filename, ILLEGAL_FILENAME_CHARS, "")
    
    -- 移除首尾空格
    sanitized = string.match(sanitized, "^%s*(.-)%s*$") or ""
    
    -- 限制长度
    if #sanitized > 255 then
        sanitized = string.sub(sanitized, 1, 255)
    end
    
    return sanitized
end

function InputSanitizer.sanitize_input(user_input)
    -- 消毒用户输入
    
    -- 移除危险字符
    local sanitized = string.gsub(user_input, "[&|<>^;]", "")
    
    -- 移除路径遍历
    sanitized = string.gsub(sanitized, "%.%.", "")
    
    -- 移除首尾空格
    sanitized = string.match(sanitized, "^%s*(.-)%s*$") or ""
    
    return sanitized
end

function InputSanitizer.escape_command_arg(arg)
    -- 转义命令行参数
    
    -- 使用引号包裹
    if string.find(arg, " ") or string.find(arg, "[&|<>^]") then
        return '"' .. arg .. '"'
    end
    
    return arg
end

-- ============================================================
-- 演示函数
-- ============================================================

local function demonstrate_input_validation()
    -- 演示输入验证
    
    print("============================================================")
    print("输入验证演示")
    print("============================================================")
    
    -- 测试用例
    local test_cases = {
        {input = "", expected = false, description = "空输入"},
        {input = "hello_world", expected = true, description = "正常输入"},
        {input = "file.txt & del /f /q C:\\*.*", expected = false, description = "命令注入攻击"},
        {input = "..\\..\\..\\Windows\\System32\\config\\SAM", expected = false, description = "路径遍历攻击"},
        {input = "file:name.txt", expected = false, description = "包含非法字符"},
        {input = "normal_file.txt", expected = true, description = "正常文件名"},
    }
    
    for _, test in ipairs(test_cases) do
        local result = SecurityValidator.validate_input(test.input)
        local status = (result == test.expected) and "PASS" or "FAIL"
        print(string.format("[%s] %s: '%s' -> %s", 
              status, test.description, test.input, tostring(result)))
    end
end

local function demonstrate_path_security()
    -- 演示路径安全
    
    print("\n============================================================")
    print("路径安全演示")
    print("============================================================")
    
    -- 创建临时目录
    local temp_dir = "temp_security_test_lua"
    os.execute(string.format('mkdir "%s" 2>nul', temp_dir))
    
    local handler = SafePathHandler.new(temp_dir)
    
    -- 测试用例
    local test_cases = {
        {input = "file.txt", expected = true, description = "正常文件"},
        {input = "..\\..\\secret.txt", expected = false, description = "路径遍历攻击"},
        {input = "C:\\Windows\\System32\\cmd.exe", expected = false, description = "绝对路径"},
        {input = "subdir\\file.txt", expected = true, description = "子目录文件"},
    }
    
    for _, test in ipairs(test_cases) do
        local result = handler:safe_join(test.input)
        local is_safe = (result ~= nil)
        local status = (is_safe == test.expected) and "PASS" or "FAIL"
        print(string.format("[%s] %s: '%s' -> %s", 
              status, test.description, test.input, is_safe and "安全" or "不安全"))
    end
    
    -- 清理临时目录
    os.execute(string.format('rmdir /s /q "%s" 2>nul', temp_dir))
end

local function demonstrate_hashing()
    -- 演示哈希计算
    
    print("\n============================================================")
    print("哈希计算演示")
    print("============================================================")
    
    local hasher = SecureHasher
    
    -- 测试字符串哈希
    local test_strings = {
        "Hello, World!",
        "Lua security",
        "Test hash",
    }
    
    for _, str in ipairs(test_strings) do
        local hash = hasher.calculate_simple_hash(str)
        print(string.format("字符串: '%s' -> 哈希: %d", str, hash))
    end
    
    -- 测试密码哈希
    print("\n密码哈希测试:")
    local password = "secure_password"
    local result = hasher.hash_password(password)
    
    print(string.format("密码: '%s'", password))
    print(string.format("盐值: '%s...'", string.sub(result.salt, 1, 16)))
    print(string.format("哈希: '%s'", result.hash))
end

local function demonstrate_logging()
    -- 演示日志记录
    
    print("\n============================================================")
    print("日志记录演示")
    print("============================================================")
    
    -- 创建临时日志目录
    local log_dir = "temp_logs_lua"
    
    local logger = AuditLogger.new(log_dir)
    
    -- 记录各种事件
    logger:log_event("INFO", "用户登录系统")
    logger:log_security_event("LOGIN_SUCCESS", "登录成功")
    logger:log_error("FILE_NOT_FOUND", "找不到文件: test.txt")
    
    print("日志记录完成")
    print(string.format("日志目录: %s", log_dir))
    
    -- 显示日志内容
    local log_file = log_dir .. "\\audit.log"
    local file = io.open(log_file, "r")
    if file then
        print("\n日志内容:")
        print(file:read("*a"))
        file:close()
    end
    
    -- 清理临时日志目录
    os.execute(string.format('rmdir /s /q "%s" 2>nul', log_dir))
end

local function demonstrate_sanitization()
    -- 演示输入消毒
    
    print("\n============================================================")
    print("输入消毒演示")
    print("============================================================")
    
    -- 测试用例
    local test_cases = {
        {input = "file:name.txt", description = "文件名消毒"},
        {input = "  hello world  ", description = "空格处理"},
        {input = "file.txt & command", description = "命令注入消毒"},
        {input = "..\\..\\path", description = "路径遍历消毒"},
    }
    
    for _, test in ipairs(test_cases) do
        local sanitized = InputSanitizer.sanitize_input(test.input)
        print(string.format("%s:", test.description))
        print(string.format("  输入: '%s'", test.input))
        print(string.format("  输出: '%s'", sanitized))
        print("")
    end
end

local function demonstrate_file_operations()
    -- 演示文件操作安全
    
    print("\n============================================================")
    print("文件操作安全演示")
    print("============================================================")
    
    -- 创建临时目录
    local temp_dir = "temp_file_test_lua"
    os.execute(string.format('mkdir "%s" 2>nul', temp_dir))
    
    local handler = SafePathHandler.new(temp_dir)
    
    -- 测试文件写入
    local test_file = "test_file.txt"
    local test_content = "Hello, this is a test file for security demonstration."
    
    print(string.format("创建测试文件: %s", test_file))
    if handler:safe_write(test_file, test_content) then
        print("  文件创建成功")
        
        -- 测试文件读取
        local content = handler:safe_read(test_file)
        if content then
            print(string.format("  文件内容: '%s'", content))
            
            -- 计算哈希
            local hash = SecureHasher.calculate_simple_hash(content)
            print(string.format("  文件哈希: %d", hash))
        else
            print("  文件读取失败")
        end
    else
        print("  文件创建失败")
    end
    
    -- 测试危险路径
    print("\n测试危险路径:")
    local dangerous_paths = {
        "..\\..\\secret.txt",
        "C:\\Windows\\System32\\cmd.exe",
        "file.txt & del /f /q C:\\*.*",
    }
    
    for _, dangerous_path in ipairs(dangerous_paths) do
        local result = handler:safe_join(dangerous_path)
        local status = (result == nil) and "PASS" or "FAIL"
        print(string.format("[%s] 危险路径被拒绝: '%s'", status, dangerous_path))
    end
    
    -- 清理临时目录
    os.execute(string.format('rmdir /s /q "%s" 2>nul', temp_dir))
end

-- ============================================================
-- 主函数
-- ============================================================

local function main()
    print("Lua安全编程对比示例")
    print("============================================================")
    
    demonstrate_input_validation()
    demonstrate_path_security()
    demonstrate_hashing()
    demonstrate_logging()
    demonstrate_sanitization()
    demonstrate_file_operations()
    
    print("\n============================================================")
    print("所有演示完成")
    print("============================================================")
end

-- 运行主函数
main()